ZenPacks/zenoss/OpenStackInfrastructure/zenopenstack.py

```diff
@@ -174,12 +178,28 @@ def set_config(self, device_id, cfg):
         # (list of OpenStackDataSourceConfig as returned by OpenStackConfig service)
 
         self.configs[device_id] = {}
+        self.component_id[device_id] = {}
+        self.resource_bytype[device_id] = defaultdict(set)
         self.event_types[device_id] = cfg.zOpenStackProcessEventTypes
 
         for datasource in cfg.datasources:
             self.configs.setdefault(device_id, {})
             self.configs[device_id].setdefault(datasource.resourceId, {})
             self.configs[device_id][datasource.resourceId][datasource.meter] = datasource.points
+            self.component_id[device_id][datasource.resourceId] = datasource.component
+            self.resource_bytype[device_id][datasource.component_meta_type].add(datasource.resourceId)
+
+    def has_resource(self, device_id, resourceId):
+        return resourceId in self.component_id.get(device_id, {})
+
+    def get_component_id(self, device_id, resourceId):
+        # Given a resource ID, return the corresponding component ID,
+        # or None if it is unrecognized.
+        return self.component_id.get(device_id, {}).get(resourceId, None)
```

Comment:
None, is the default. not needed.

Reply:
True.  I have a habit of doing that from time to time even when i know it's not needed.  I guess I just like the explicit return value.